src/ScheduleFunctions.cpp

```diff
@@ -1518,103 +1380,128 @@ class InjectGroupRealization : public IRMutator2 {
+        for (size_t i = add_lets.size(); i > 0; --i) {
+            const auto &b = add_lets[i - 1];
+            producer = LetStmt::make(b.first, b.second, producer);
         }
 
-        if (store_level.match(for_loop->name)) {
-            debug(3) << "Found store level at " << for_loop->name << "\n";
-            internal_assert(found_compute_level)
-                << "The compute loop level was not found within the store loop level!\n";
-            body = build_realize_group(body);
-            found_store_level = true;
+        // The original bounds of the loop nests (without any loop-fusion)
+        auto bounds = CollectBounds::collect_bounds(producer);
+
+        // Compute the shift factors based on the alignment strategies
+        // starting from the the parent (root loop) to the children. The root
+        // loop bounds should remain unchanged.
+        map<string, Expr> shifts;
+        for (auto i = funcs.size(); i-- > 0;) {
+            const auto& func = funcs[i];
+            if (!skip[func.name()]) {
+                compute_shift_factor(skip, func, func.name() + ".s0.", func.definition(), bounds, shifts);
+                for (size_t j = 0; j < func.updates().size(); ++j) {
+                    string prefix = func.name() + ".s" + std::to_string(j + 1) + ".";
+                    compute_shift_factor(skip, func, prefix, func.updates()[j], bounds, shifts);
+                }
+            }
         }
+        // Shift the loops.
+        producer = ShiftLoopNest::apply_shift(shifts, producer);
 
-        // Reinstate the let statements
-        for (size_t i = lets.size(); i > 0; i--) {
-            body = LetStmt::make(lets[i - 1].first, lets[i - 1].second, body);
+        // Replace all the child fused loops with the appropriate bounds
+        // (i.e. the min/max should refer to the parent loop vars and the
+        // extent should be one).
+        producer = substitute_fused_bounds(producer, replacements);
+
+        // Replace the bounds of parent fused loop with union of bounds of
+        // the fused loops.
+        producer = replace_parent_bound_with_union_bound(skip, funcs[parent_index], producer, bounds);
+
+        // Add the producer nodes.
+        for (const auto &i : funcs) {
+            if (!skip[i.name()]) {
+                producer = ProducerConsumer::make_produce(i.name(), producer);
+            }
         }
 
-        // Reinstate the placeholder prefetches
-        for (size_t i = placeholder_prefetches.size(); i > 0; i--) {
-            body = Prefetch::make(std::get<0>(placeholder_prefetches[i - 1]),
-                                  std::get<1>(placeholder_prefetches[i - 1]),
-                                  Region(),
-                                  std::get<2>(placeholder_prefetches[i - 1]),
-                                  const_true(),
-                                  body);
+        // Add the consumer nodes.
+        for (auto i = 0u; i < funcs.size(); i++) {
```

Comment:
More idiomatic: size_t i = 0

Reply:
to expand: generally, only use auto where it removes redundancy, or the type is both verbose and obvious.

Reply:
gotcha